crates/katana/core/src/starknet/transaction.rs

```diff
@@ -108,38 +108,43 @@ impl StarknetTransaction {
     pub fn l2_to_l1_messages(&self) -> Vec<MessageToL1> {
         let mut messages: Vec<MessageToL1> = vec![];
 
-        let Some(ref execution_info) = self.execution_info else {
-            return messages;
-        };
+        fn get_messages_recursively(info: &CallInfo) -> Vec<MessageToL1> {
+            let mut messages: Vec<MessageToL1> = vec![];
 
-        if let Some(ref info) = execution_info.validate_call_info {
             messages.extend(info.execution.l2_to_l1_messages.iter().map(|m| MessageToL1 {
                 to_address: m.message.to_address,
                 payload: m.message.payload.clone(),
                 from_address: info.call.caller_address,
-            }))
+            }));
+
+            info.inner_calls.iter().for_each(|call| {
+                messages.extend(get_messages_recursively(call));
+            });
+
+            messages
+        }
+
+        let Some(ref execution_info) = self.execution_info else {
+            return messages;
+        };
+
+        if let Some(ref info) = execution_info.validate_call_info {
+            messages.extend(get_messages_recursively(info));
         }
 
         if let Some(ref info) = execution_info.execute_call_info {
-            messages.extend(info.execution.l2_to_l1_messages.iter().map(|m| MessageToL1 {
-                to_address: m.message.to_address,
-                payload: m.message.payload.clone(),
-                from_address: info.call.caller_address,
-            }))
+            messages.extend(get_messages_recursively(info));
         }
 
         if let Some(ref info) = execution_info.fee_transfer_call_info {
-            messages.extend(info.execution.l2_to_l1_messages.iter().map(|m| MessageToL1 {
-                to_address: m.message.to_address,
-                payload: m.message.payload.clone(),
-                from_address: info.call.caller_address,
-            }))
+            messages.extend(get_messages_recursively(info));
         }
 
         messages
     }
 
-    fn output(&self) -> TransactionOutput {
+
+    pub(crate) fn output(&self) -> TransactionOutput {
```

Comment:
@kariy @tarrencev I am not sure this is the best rusty way to do that. But it was the one I see with minimal changes.

Please let me know if there is a better alternative, I will be happy to know it and change to it. :+1: